coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/CoralSqlNodeToTrinoSqlNodeConverter.java

```diff
@@ -0,0 +1,235 @@
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.util.SqlShuttle;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+
+import com.linkedin.coral.com.google.common.collect.ImmutableList;
+import com.linkedin.coral.common.functions.CoralSqlUnnestOperator;
+import com.linkedin.coral.trino.rel2trino.functions.TrinoArrayTransformFunction;
+
+import static org.apache.calcite.rel.rel2sql.SqlImplementor.*;
+import static org.apache.calcite.sql.parser.SqlParserPos.*;
+
+
+/**
+ * CoralSqlNodeToTrinoSqlNodeConverter rewrites the Coral SqlNode AST. It replaces Coral IR SqlCalls
+ * with Trino compatible SqlCalls to subsequently obtain a Trino compatible SqlNode AST representation.
+ * This will enable generating a SQL which can be accurately interpreted by the Trino engine.
+ *
+ * This is achieved by visiting the Coral SqlNode AST in a pre-order traversal manner and
+ * transforming each SqlNode (SqlCall), wherever required.
+ * The transformation may involve change in operator, reordering the operands
+ * or even re-constructing the SqlCall.
+ */
+public class CoralSqlNodeToTrinoSqlNodeConverter extends SqlShuttle {
+
+  public CoralSqlNodeToTrinoSqlNodeConverter() {
+  }
+
+  @Override
+  public SqlNode visit(final SqlCall call) {
+    SqlCall transformedSqlCall = getTransformedSqlCall(call);
+    return super.visit(transformedSqlCall);
+  }
+
+  public static SqlCall getTransformedSqlCall(SqlCall sqlCall) {
+    switch (sqlCall.getOperator().kind) {
+      case SELECT:
+        return getTransformedSqlSelectSqlCall(sqlCall);
+      case JOIN:
+        return getTransformedJoinSqlCall(sqlCall);
+      case AS:
+        return getTransformedAsSqlCall(sqlCall);
+      case UNNEST:
+        return getTransformedUnnestSqlCall(sqlCall);
+      case EQUALS:
+        return getTransformedEqualsOperatorSqlCall(sqlCall);
+      default:
+        return sqlCall;
+    }
+  }
+
+  // Append TryCast operator to both operands to cast each operand's data type to VARCHAR
+  private static SqlCall getTransformedEqualsOperatorSqlCall(SqlCall sqlCall) {
+    List<SqlNode> updatedOperands = new ArrayList<>();
+
+    final SqlTypeNameSpec varcharTypeNameSpec = new SqlBasicTypeNameSpec(SqlTypeName.VARCHAR, ZERO);
+    SqlDataTypeSpec varcharSqlDataTypeSpec = new SqlDataTypeSpec(varcharTypeNameSpec, ZERO);
+
+    for (SqlNode operand : sqlCall.getOperandList()) {
+      SqlNode newOperand = TrinoTryCastFunction.INSTANCE.createCall(POS,
+          new ArrayList<>(Arrays.asList(operand, varcharSqlDataTypeSpec)));
+      updatedOperands.add(newOperand);
+    }
+
+    return sqlCall.getOperator().createCall(POS, updatedOperands);
+  }
+
+  // Update unnest operand for trino engine to expand the unnest operand to a single column
+  private static SqlCall getTransformedUnnestSqlCall(SqlCall sqlCall) {
```

Comment:
Looks like the `SUPPORT_LEGACY_UNNEST_ARRAY_OF_STRUCT` is not used, which may cause regression for LI internal use? See #158 for more details.